pkg/provider/azure_loadbalancer.go

```diff
@@ -1658,16 +1771,104 @@ func (az *Cloud) reconcileLoadBalancer(clusterName string, service *v1.Service,
 		defer func() {
 			_ = az.lbCache.Delete(lbName)
 		}()
-		err := az.VMSet.EnsureHostsInPool(service, nodes, lbBackendPoolID, vmSetName, isInternal)
-		if err != nil {
-			return nil, err
+
+		if strings.EqualFold(az.Config.LoadBalancerBackendPoolConfigurationType, consts.LoadBalancerBackendPoolConfigurationTypeVMNIC) {
+			err := az.VMSet.EnsureHostsInPool(service, nodes, lbBackendPoolID, vmSetName, isInternal)
+			if err != nil {
+				return nil, err
+			}
+		} else if strings.EqualFold(az.LoadBalancerBackendPoolConfigurationType, consts.LoadBalancerBackendPoolConfigurationTypeVMIP) {
+			backendPools := *lb.BackendAddressPools
+			if err := az.ensureNodeIPAddressesInPool(service, clusterName, lbName, backendPools, nodes); err != nil {
+				return nil, fmt.Errorf("reconcileLoadBalancer for service(%s): failed to ensure node IP addresses in pool: %w", serviceName, err)
+			}
 		}
 	}
 
 	klog.V(2).Infof("reconcileLoadBalancer for service(%s): lb(%s) finished", serviceName, lbName)
 	return lb, nil
 }
 
+func (az *Cloud) ensureNodeIPAddressesInPool(service *v1.Service, clusterName, lbName string, backendPools []network.BackendAddressPool, nodes []*v1.Node) error {
+	vnetResourceGroup := az.ResourceGroup
+	if len(az.VnetResourceGroup) > 0 {
+		vnetResourceGroup = az.VnetResourceGroup
+	}
+	vnetID := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s", az.SubscriptionID, vnetResourceGroup, az.VnetName)
+
+	changed := false
+	lbBackendPoolName := getBackendPoolName(clusterName, service)
+	var managedBackendPool network.BackendAddressPool
+	for i, backendPool := range backendPools {
+		if strings.EqualFold(to.String(backendPool.Name), lbBackendPoolName) &&
+			backendPool.BackendAddressPoolPropertiesFormat != nil {
+			if backendPool.LoadBalancerBackendAddresses == nil {
+				lbBackendPoolAddresses := make([]network.LoadBalancerBackendAddress, 0)
+				backendPool.LoadBalancerBackendAddresses = &lbBackendPoolAddresses
+			}
+
+			existingIPs := sets.NewString()
+			for _, loadBalancerBackendAddress := range *backendPool.LoadBalancerBackendAddresses {
+				if loadBalancerBackendAddress.LoadBalancerBackendAddressPropertiesFormat != nil &&
+					loadBalancerBackendAddress.IPAddress != nil {
+					klog.V(4).Infof("ensureNodeIPAddressesInPool: found existing IP %s in the backend pool %s", to.String(loadBalancerBackendAddress.IPAddress), lbBackendPoolName)
+					existingIPs.Insert(to.String(loadBalancerBackendAddress.IPAddress))
+				}
+			}
+
+			for _, node := range nodes {
+				if isControlPlaneNode(node) {
+					klog.V(4).Infof("ensureNodeIPAddressesInPool: skipping control plane node %s", node.Name)
+					continue
+				}
+
+				useMultipleSLBs := strings.EqualFold(az.LoadBalancerSku, "standard") && az.EnableMultipleStandardLoadBalancers
+				if strings.EqualFold(az.LoadBalancerSku, "basic") || useMultipleSLBs {
+					vmSetName, err := az.VMSet.GetNodeVMSetName(node)
```

Comment:
could you skip the standalone VM for this case? (and same for other places)

Reply:
The standalone VM will be skipped because az.VMSet.GetNodeVMSetName will return an empty string.